### PR TITLE
Numbers based TD

### PR DESCRIPTION
**Pull request by clee2000**
Convert from a list/bucket based TD system to just a numbers based TD system.  Looks like a massive change but a decent amount of it is tests and removing code.

Main file of interest is interface.py, which Github is collapsing by default due to size

The test files pretty much got rewritten entirely since a lot of the old tests are no longer relevant.

Other notable changes:
* Use Frozenset to make TestRun hashable
* Adds tools/test/heuristics/__init__.py to ensure that unittest can discover the tests

### CODE REVIEW

File: .github/workflows/pull.yml [hunk 1]
```diff
@@ -46,397 +46,3 @@ jobs:
       build-environment: linux-jammy-py3.8-gcc11
       docker-image: ${{ needs.linux-jammy-py3_8-gcc11-build.outputs.docker-image }}
       test-matrix: ${{ needs.linux-jammy-py3_8-gcc11-build.outputs.test-matrix }}
-
-  linux-docs:
```
Comment (osalpekar) [inline .github/workflows/pull.yml:50]: q: are the jobs removed here just to run less CI on this PR?

**Review by osalpekar — APPROVED** [review]

File: .github/workflows/pull.yml (hunk 1 above)
Comment (clee2000) [reply .github/workflows/pull.yml:50]: yes, I just want a sanity check most of the time

**Review by huydhn — APPROVED** [review]
The testing looks good, so let's land it :)

**Comment by clee2000** [comment]
@pytorch merge -f "get around sanity check"

**Pull request closed without merging**